### ISSUE TITLE
Battery block crashing without a battery even with `allow_missing = true`

### ISSUE THREAD

**Issue by LordMZTE** [greshake/i3status-rust#1424]
It seems like commit c68247 introduced a bug that causes this line
https://github.com/greshake/i3status-rust/blob/a5c1bc6d95594841e0e5ad5a2ffc97a214df0427/src/blocks/battery.rs#L818
to be reached, even if `allow_missing` is set to true in the config.

Ideally, nothing should happen here, but this actually causes an error.

**Comment by tarkh** [comment]
Same here on Arch

**Comment by brownjohnf** [comment]
Naively, it seems like `device_str` (https://github.com/greshake/i3status-rust/blob/master/src/blocks/battery.rs#L797) should be an `Option<String>`, and the various `BatteryDriver`s updated to accept an `Option<String>` for their device strings. Alternatively, you could just short-circuit and have a dummy `BatteryDriver::Missing` that gets used if the `device_str` can't be determined and `allow_missing` is true.

I've never really looked at this code before though, and I'm not sure what's idiomatic for the blocks and drivers.

**Comment by brownjohnf** [comment]
(And I'm having the same issue, also on Arch.)

**Comment by ammgws** [comment]
OK so I believe d6fbfd06cc4d078efccb1c559e7eb934d36ffe7a should resolve this issue by restoring previous behavior for users whose battery shows up as BAT0 by default. In all other cases where the battery shows up as something else, such as CMB1 for my laptop, the block will not error out if you have `allow_missing`, but it also won't update if your battery shows up again (however I don't think we ever had this working), so you would need to explicitly name the battery in your config. This is a similar issue to the one discussed in https://github.com/greshake/i3status-rust/issues/1362, though that one was mainly about upower.

**Comment by ammgws** [comment]
@MaxVerevkin Is there anything else we can do apart from properly implementing device refresh logic?

**Comment by MaxVerevkin** [comment]
I mean, if this is how it was implemented before, then I guess it's fine.

**Comment by ammgws** [comment]
Fixed in 0.21.5

**Comment by tarkh** [comment]
FYI, when I upgraded to the latest version (I'm on Arch @ Macbook Pro), i3status-rust did not start, giving an error:
```
error in block 'battery': failed to open file /sys/class/power_supply/apple_mfi_fastcharge/status
```
So I had to add an option `driver: "upower"` to config, and all went good. Also setting this option solves #1410 for me.
